File: backend-python-legacy/app/refresh_cache.py
"""TTL por fonte + last-good. O hub continua a 60 s; cada API só é batida quando vence.

SSE e `/display`/firmware não mudam: o snapshot vai a cada `USAGE_INTERVAL_S`.
O que muda é *quando* o coletor gasta a cota de um terceiro (CoinGecko, AdSense, clima).
"""

from __future__ import annotations

import json
import threading
import time
from typing import Any

from app.http_util import is_rate_limit, result_is_rate_limited, retry_after_s
from app.store import provider as provider_cfg

# Intervalo mínimo entre chamadas *reais*. Cotas de assinatura acompanham o hub.
# Mercado/clima mudam devagar e as APIs públicas 429am se marteladas no mesmo ritmo.
TTL_S: dict[str, int] = {
    "claude": 60,
    "gpt": 60,
    "cursor": 60,
    "openrouter": 60,
    "deepseek": 60,
    "opencode": 60,
    "fal": 60,
    "bitcoin": 60,  # Blockstream; a cotação CoinGecko tem TTL próprio (5 min)
    "adsense": 300,
    "weather": 600,
    "currencies": 60,  # forex/CoinGecko têm TTL próprio dentro do provedor
}

# GET /usage força cotas na hora. Mercado/clima nunca — 429 não se resolve martelando.
FORCEABLE = frozenset(
    {
        "claude",
        "gpt",
        "cursor",
        "openrouter",
        "deepseek",
        "opencode",
        "fal",
    }
)

_DEFAULT_TTL_S = 60
_MIN_BACKOFF_S = 60.0


class RefreshCache:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._fresh_at: dict[str, float] = {}
        self._value: dict[str, Any] = {}
        self._fingerprint: dict[str, str] = {}
        self._backoff_until: dict[str, float] = {}

    def reset(self) -> None:
        with self._lock:
            self._fresh_at.clear()
            self._value.clear()
            self._fingerprint.clear()
            self._backoff_until.clear()

    def get(self, name: str) -> Any | None:
        with self._lock:
            return self._value.get(name)

    def due(self, name: str, *, fingerprint: str = "", force: bool = False) -> bool:
        now = time.monotonic()
        ttl = TTL_S.get(name, _DEFAULT_TTL_S)
        with self._lock:
            if now < self._backoff_until.get(name, 0.0):
                return False
            if self._fingerprint.get(name) != fingerprint:
                return True
            if force and name in FORCEABLE:
                return True
            last = self._fresh_at.get(name)
            if last is None:
                return True
            return (now - last) >= ttl

    def store(self, name: str, value: Any, *, fingerprint: str = "") -> None:
        with self._lock:
            self._value[name] = value
            self._fresh_at[name] = time.monotonic()
            self._fingerprint[name] = fingerprint
            self._backoff_until.pop(name, None)

    def note_rate_limit(self, name: str, retry_after: float | None, value: Any | None = None) -> None:
        ttl = float(TTL_S.get(name, _DEFAULT_TTL_S))
        wait = max(_MIN_BACKOFF_S, retry_after or ttl)
        with self._lock:
            self._backoff_until[name] = time.monotonic() + wait
            if value is not None and name not in self._value:
                self._value[name] = value

    def take(self, name: str, value: Any, *, fingerprint: str = "", error: BaseException | str | None = None) -> Any:
        """Guarda sucesso, ou last-good em 429; devolve o que o payload deve usar."""
        limited = is_rate_limit(error) or result_is_rate_limited(value)
        if limited:
            previous = self.get(name)
            self.note_rate_limit(name, retry_after_s(error or value), value=value)
            return previous if previous is not None else value
        self.store(name, value, fingerprint=fingerprint)
        return value


def fingerprint(cfg: dict[str, Any], name: str) -> str:
    """Muda quando a config da fonte muda — evita servir cache de outra carteira/cidade."""
    if name in ("weather", "currencies"):
        blob: Any = cfg.get(name) or {}
    else:
        blob = provider_cfg(cfg, name)
    try:
        return json.dumps(blob, sort_keys=True, default=str)
    except TypeError:
        return str(blob)


cache = RefreshCache()

File: backend/src/refreshCache.ts
import { isRateLimit, retryAfterS, resultIsRateLimited } from "./httpClient.js";
import { provider as providerCfg } from "./store.js";

export const TTL_S: Record<string, number> = {
  claude: 60,
  gpt: 60,
  cursor: 60,
  openrouter: 60,
  deepseek: 60,
  opencode: 60,
  fal: 60,
  bitcoin: 60,
  adsense: 300,
  weather: 600,
  currencies: 60,
};

export const FORCEABLE: ReadonlySet<string> = new Set([
  "claude",
  "gpt",
  "cursor",
  "openrouter",
  "deepseek",
  "opencode",
  "fal",
]);

const DEFAULT_TTL_S = 60;
const MIN_BACKOFF_S = 60.0;

function nowMonotonic(): number {
  // performance.now is monotonic (ms), convert to seconds
  return performance.now() / 1000;
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) return String(value);
  if (typeof value !== "object") return String(value);
  if (Array.isArray(value)) {
    return "[" + value.map((v) => stableStringify(v)).join(",") + "]";
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  const parts: string[] = [];
  for (const k of keys) {
    let v: string;
    try {
      const raw = obj[k];
      if (raw !== null && typeof raw === "object") v = stableStringify(raw);
      else if (typeof raw === "string") v = JSON.stringify(raw);
      else if (typeof raw === "number" || typeof raw === "boolean") v = JSON.stringify(raw);
      else if (raw === null || raw === undefined) v = "null";
      else v = JSON.stringify(String(raw));
    } catch {
      v = JSON.stringify(String(obj[k]));
    }
    parts.push(JSON.stringify(k) + ":" + v);
  }
  return "{" + parts.join(",") + "}";
}

export class RefreshCache {
  private freshAt = new Map<string, number>();
  private value = new Map<string, unknown>();
  private fingerprint = new Map<string, string>();
  private backoffUntil = new Map<string, number>();
  private inflight = new Map<string, Promise<unknown>>();

  reset(): void {
    this.freshAt.clear();
    this.value.clear();
    this.fingerprint.clear();
    this.backoffUntil.clear();
    this.inflight.clear();
  }

  get(name: string): unknown | null {
    return this.value.has(name) ? this.value.get(name)! : null;
  }

  due(name: string, opts: { fingerprint?: string; force?: boolean } = {}): boolean {
    const fingerprint = opts.fingerprint ?? "";
    const force = opts.force ?? false;
    const now = nowMonotonic();
    const ttl = TTL_S[name] ?? DEFAULT_TTL_S;
    if (now < (this.backoffUntil.get(name) ?? 0)) return false;
    if ((this.fingerprint.get(name) ?? null) !== fingerprint) return true;
    if (force && FORCEABLE.has(name)) return true;
    const last = this.freshAt.get(name);
    if (last === undefined) return true;
    return now - last >= ttl;
  }

  store(name: string, value: unknown, opts: { fingerprint?: string } = {}): void {
    const fingerprint = opts.fingerprint ?? "";
    this.value.set(name, value);
    this.freshAt.set(name, nowMonotonic());
    this.fingerprint.set(name, fingerprint);
    this.backoffUntil.delete(name);
  }

  noteRateLimit(name: string, retryAfter: number | null | undefined, value: unknown | null = null): void {
    const ttl = Number(TTL_S[name] ?? DEFAULT_TTL_S);
    const wait = Math.max(MIN_BACKOFF_S, retryAfter ?? ttl);
    this.backoffUntil.set(name, nowMonotonic() + wait);
    if (value !== null && value !== undefined && !this.value.has(name)) {
      this.value.set(name, value);
    }
  }

  take(
    name: string,
    value: unknown,
    opts: { fingerprint?: string; error?: unknown } = {},
  ): unknown {
    const fingerprint = opts.fingerprint ?? "";
    const error = opts.error ?? null;
    const limited = isRateLimit(error) || resultIsRateLimited(value);
    if (limited) {
      const previous = this.get(name);
      this.noteRateLimit(name, retryAfterS((error ?? value) as unknown), value);
      return previous !== null && previous !== undefined ? previous : value;
    }
    this.store(name, value, { fingerprint });
    return value;
  }

  // Inflight coalescing: ensure concurrent fetches for same name share one promise
  async coalesce<T>(name: string, fn: () => Promise<T>): Promise<T> {
    const existing = this.inflight.get(name);
    if (existing) return existing as Promise<T>;
    const p = (async () => {
      try {
        return await fn();
      } finally {
        this.inflight.delete(name);
      }
    })();
    this.inflight.set(name, p as Promise<unknown>);
    return p;
  }

  // Convenience: check due and coalesce fetch with caching
  async getOrFetch<T>(
    name: string,
    fingerprint: string,
    fetcher: () => Promise<T>,
    opts: { force?: boolean } = {},
  ): Promise<T> {
    if (!this.due(name, { fingerprint, force: opts.force ?? false })) {
      return this.get(name) as T;
    }
    return this.coalesce(name, async () => {
      // re-check due inside coalesce gate to avoid race
      if (!this.due(name, { fingerprint, force: opts.force ?? false })) {
        return this.get(name) as T;
      }
      try {
        const value = await fetcher();
        this.take(name, value, { fingerprint });
        return value as T;
      } catch (e) {
        if (isRateLimit(e)) {
          const prev = this.get(name);
          this.noteRateLimit(name, retryAfterS(e), null);
          if (prev !== null) return prev as T;
        }
        throw e;
      }
    });
  }
}

export function fingerprint(cfg: Record<string, unknown>, name: string): string {
  let blob: unknown;
  if (name === "weather" || name === "currencies") {
    blob = (cfg as Record<string, unknown>)[name] ?? {};
  } else {
    blob = providerCfg(cfg as Record<string, unknown>, name);
  }
  try {
    // Use stable sorted JSON
    if (blob !== null && typeof blob === "object") return stableStringify(blob);
    return JSON.stringify(blob, null, 0) ?? String(blob);
  } catch {
    return String(blob);
  }
}

export const cache = new RefreshCache();

// snake_case aliases
export const TTL_s = TTL_S;
export const FORCEABLE_SET = FORCEABLE;
